### PR TITLE
Nuke - Lockdrop and LBA - Transfer maUST / remove astroport liquidity and transfer funds to the user

### PR DESCRIPTION
**Pull request by alienHunterOnMars**

### CODE REVIEW

File: contracts/auction/Cargo.toml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "mars-auction"
-version = "1.0.0"
+version = "1.0.2"
```
Comment (spikeonmars) [inline contracts/auction/Cargo.toml:3]: If we want to be correct about this and if I understood correctly. This seems like a minor release (1.1.0) and not a patch release, as we are adding features/changing the api without breaking backwards compatibility

**Review by spikeonmars — APPROVED** [review]

File: contracts/lockdrop/src/contract.rs [hunk 2]
```diff
@@ -655,6 +656,121 @@ pub fn handle_deposit_mars_to_auction(
...
+            let (user_bytes, position) = item?;
+            let user = String::from_utf8(user_bytes)?;
+            Ok((user, position))
+        })
+        .collect::<StdResult<Vec<_>>>()?;
+
+    // Loop over user positions
+    for (user_addr, _) in user_list.iter() {
+        // USER INFO :: RETRIEVE --> UPDATE
+        let mut user_info = USER_INFO
```
Comment (larry0x) [inline contracts/lockdrop/src/contract.rs:686]: i'm not super familiar with how lockdrop contract works, but why don't you simply do
```rust
for (user_addr, user_info) in user_list { .. }
```
here?

File: contracts/lockdrop/src/contract.rs [hunk 3]
```diff
@@ -655,6 +656,121 @@ pub fn handle_deposit_mars_to_auction(
...
+
+            if let Some(index) = index_search {
+                user_info.lockup_positions.remove(index);
+            } else {
+                return Err(StdError::generic_err(format!(
+                    "Lockup position not found for id {}",
+                    unlocked_lockup_id
+                )));
+            }
+        }
```
Comment (larry0x) [inline contracts/lockdrop/src/contract.rs:759]: i think you can do smth like
```rust
// cast ids into a hashset. optional, will save gas
let unlocked_ids: HashSet<_> = lockup_positions_unlocked.into_iter().collect();
user_info.lockup_positions.retain(|id| !unlocked_ids.contains(id));
```

File: packages/mars-periphery/src/lockdrop.rs [hunk 4]
```diff
@@ -64,6 +64,8 @@ pub enum ExecuteMsg {
     },
     /// Called by the bootstrap auction contract when liquidity is added to the MARS-UST Pool to enable MARS withdrawals by users
     EnableClaims {},
+    /// Loops over provided user addresses, unlocks each of their lockup position. Rmoves the list of user positions from state
```
Comment (dancreee) [inline packages/mars-periphery/src/lockdrop.rs:67]: Rmoves -> Removes

File: contracts/lockdrop/src/contract.rs [hunk 5]
```diff
@@ -655,6 +656,121 @@ pub fn handle_deposit_mars_to_auction(
...
+            // UPDATE STATE
+            state.total_maust_locked -= maust_to_withdraw;
+
+            total_ust_unlocked = total_ust_unlocked.checked_add(lockup_info.ust_locked)?;
+            total_maust_withdrawn = total_maust_withdrawn.checked_add(maust_to_withdraw)?;
+            lockup_positions_unlocked.push(lockup_id.clone());
+
+            // DISSOLVE LOCKUP POSITION
+            lockup_info.ust_locked = Uint128::zero();
+            // remove_lockup_pos_from_user_info(&mut user_info, lockup_id.clone())?;
```
Comment (dancreee) [inline contracts/lockdrop/src/contract.rs:720]: Can remove this comment? - could have used this func below on ln #746 too, instead of reimplementing it

**Review by dancreee — APPROVED** [review]
LGTM, some minor comments

File: contracts/auction/Cargo.toml (hunk 1 above)
Comment (alienHunterOnMars) [reply contracts/auction/Cargo.toml]: resolved

File: contracts/lockdrop/src/contract.rs [hunk 6]
```diff
@@ -655,6 +656,121 @@ pub fn handle_deposit_mars_to_auction(
...
+    // Loop over user positions
+    for (user_addr, _) in user_list.iter() {
+        // USER INFO :: RETRIEVE --> UPDATE
+        let mut user_info = USER_INFO
+            .may_load(deps.storage, &Addr::unchecked(user_addr.clone()))?
+            .unwrap_or_default();
+
+        // Remove userInfo if all UST has been refunded
+        if user_info.total_ust_locked == Uint128::zero() {
+            USER_INFO.remove(deps.storage, &Addr::unchecked(user_addr));
```
[the remark this answers is not in the document]
Comment (alienHunterOnMars) [reply contracts/lockdrop/src/contract.rs:692]: resolved

File: contracts/lockdrop/src/contract.rs (hunk 3 above)
Comment (alienHunterOnMars) [reply contracts/lockdrop/src/contract.rs]: resolved

File: contracts/lockdrop/src/contract.rs [hunk 7]
```diff
@@ -655,6 +656,121 @@ pub fn handle_deposit_mars_to_auction(
...
+            } else {
+                return Err(StdError::generic_err(format!(
+                    "Lockup position not found for id {}",
+                    unlocked_lockup_id
+                )));
+            }
+        }
+
+        // Remove userInfo if all UST has been refunded
+        if user_info.total_ust_locked == Uint128::zero() {
```
[the remark this answers is not in the document]
Comment (alienHunterOnMars) [reply contracts/lockdrop/src/contract.rs]: resolved

File: packages/mars-periphery/src/lockdrop.rs (hunk 4 above)
Comment (alienHunterOnMars) [reply packages/mars-periphery/src/lockdrop.rs]: resolved

File: contracts/lockdrop/src/contract.rs (hunk 5 above)
Comment (alienHunterOnMars) [reply contracts/lockdrop/src/contract.rs]: resolved

**Review by dancreee — APPROVED** [review]

**Pull request merged**